File: Data Structures & Algorithms/binary-tree-diameter/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
   
 
        result = 0 

        def helper(root): 
            nonlocal result     
          
            if root:
                left = helper(root.left)
                right = helper(root.right)
                result = max(result, left + right)
                return 1 + max(left, right)
            else:
                return 0

        helper(root)
        return result

        



    
        

File: Data Structures & Algorithms/pow-x-n/submission-3.cpp
class Solution {
private:
// x=2.00000
// n=5

// helper(3) * helper(3)
// helper(1) * helper(1) * helper(1) * helper(1)
// 

    double helper(double num, long long n){
        if (n == 0){
            return 1;
        }
      
        double power = helper(num, n / 2);
        double result = power * power;
        if (n % 2 == 1){
            result *= num; 
        }

        return result;
    }

public:
    double myPow(double x, int n) {
        
        long long num = n;

        double result;
        result = helper(x, std::abs(num));
        
        std::cout << result << std::endl;

        if (num < 0){
            return 1.0 / result;
        }
        return result;
    }
};
